src/query/execute_statement.c

```diff
@@ -14809,12 +14809,67 @@ do_replicate_statement (PARSER_CONTEXT * parser, PT_NODE * statement)
     {
       PT_PRINT_VALUE_FUNC saved_func = parser->print_db_value;
       int saved_custom_print = parser->custom_print;
+      PARSER_VARCHAR **host_val;
+      char *sbr_text, *sql_text;
+      int i, n, nth;
+      int sql_len = strlen (parser->context.sql_user_text);
+      int var_len = 0;
+      bool begin_quote = false;
+
+      host_val = (PARSER_VARCHAR **) malloc (sizeof (PARSER_VARCHAR *) * parser->host_var_count);
+      if (host_val == NULL)
+	{
+	  return ER_OUT_OF_VIRTUAL_MEMORY;
+	}
+
+      for (i = 0; i < parser->host_var_count; i++)
+	{
+	  host_val[i] = pt_print_db_value (parser, &parser->host_variables[i]);
+	  var_len += host_val[i]->length;
+	}
+
+      sbr_text = (char *) malloc (sql_len + var_len);
+      if (sbr_text == NULL)
+	{
+	  return ER_OUT_OF_VIRTUAL_MEMORY;
+	}
 
       parser->custom_print |= (PT_PRINT_ORIGINAL_BEFORE_CONST_FOLDING | PT_PRINT_QUOTES);
       parser->print_db_value = pt_print_node_value;
-      repl_stmt.stmt_text = parser_print_tree (parser, statement);
+
+      n = nth = 0;
+
+      sql_text = parser->context.sql_user_text;
+      for (i = 0; i < sql_len; i++)
+	{
+	  if (sql_text[i] == '\'')
+	    {
+	      if (!begin_quote)
+		{
+		  begin_quote = true;
+		}
+	      else
+		{
+		  begin_quote = false;
+		}
+	    }
+	  if (sql_text[i] == '?' && !begin_quote)
+	    {
+	      strncpy (&sbr_text[n], (char *) host_val[nth]->bytes, host_val[nth]->length);
+	      n += host_val[nth++]->length;
+	    }
+	  else
+	    {
+	      sbr_text[n++] = sql_text[i];
+	    }
+	}
+
+      sbr_text[n] = 0;
+      repl_stmt.stmt_text = sbr_text;
       parser->print_db_value = saved_func;
       parser->custom_print = saved_custom_print;
+
+      free (host_val);
```

Comment:
Where is the sbr_text free?

Reply:
sbr_text should be free after the SBR's query string is used, so it must not be free.

Reply:
Do you mean the sbr_text is deallocation somewhere? I can't find where it is deallocation.

Reply:
I fixed to free all of allocated memories at the end of the function do_replicate_statement.